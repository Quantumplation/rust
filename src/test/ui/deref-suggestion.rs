macro_rules! borrow {
    ($x:expr) => { &$x } //~ ERROR arguments to this function are incorrect
}

fn foo(_: String) {}

fn foo2(s: &String) {
    foo(s);
    //~^ ERROR arguments to this function are incorrect
}

fn foo3(_: u32) {}
fn foo4(u: &u32) {
    foo3(u);
    //~^ ERROR arguments to this function are incorrect
}

struct S<'a> {
    u: &'a u32,
}

struct R {
    i: u32,
}

fn main() {
    let s = String::new();
    let r_s = &s;
    foo2(r_s);
    foo(&"aaa".to_owned());
    //~^ ERROR arguments to this function are incorrect
    foo(&mut "aaa".to_owned());
    //~^ ERROR arguments to this function are incorrect
    foo3(borrow!(0));
    foo4(&0);
    assert_eq!(3i32, &3i32);
    //~^ ERROR arguments to this function are incorrect
    let u = 3;
    let s = S { u };
    //~^ ERROR mismatched types
    let s = S { u: u };
    //~^ ERROR mismatched types
    let i = &4;
    let r = R { i };
    //~^ ERROR mismatched types
    let r = R { i: i };
    //~^ ERROR mismatched types
}
