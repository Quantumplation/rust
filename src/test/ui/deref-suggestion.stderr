error[E0308]: arguments to this function are incorrect
  --> $DIR/deref-suggestion.rs:8:5
   |
LL |     foo(s);
   |     ^^^^-^
   |         |
   |         expected `String`, found `&String`
   |
note: function defined here
  --> $DIR/deref-suggestion.rs:5:4
   |
LL | fn foo(_: String) {}
   |    ^^^ ---------
help: provide an argument of the correct type
   |
LL |     foo({String});
   |     ^^^^^^^^^^^^^

error[E0308]: arguments to this function are incorrect
  --> $DIR/deref-suggestion.rs:14:5
   |
LL |     foo3(u);
   |     ^^^^^-^
   |          |
   |          expected `u32`, found `&u32`
   |
note: function defined here
  --> $DIR/deref-suggestion.rs:12:4
   |
LL | fn foo3(_: u32) {}
   |    ^^^^ ------
help: provide an argument of the correct type
   |
LL |     foo3({u32});
   |     ^^^^^^^^^^^

error[E0308]: arguments to this function are incorrect
  --> $DIR/deref-suggestion.rs:30:5
   |
LL |     foo(&"aaa".to_owned());
   |     ^^^^-----------------^
   |         |
   |         expected `String`, found `&String`
   |
note: function defined here
  --> $DIR/deref-suggestion.rs:5:4
   |
LL | fn foo(_: String) {}
   |    ^^^ ---------
help: provide an argument of the correct type
   |
LL |     foo({String});
   |     ^^^^^^^^^^^^^

error[E0308]: arguments to this function are incorrect
  --> $DIR/deref-suggestion.rs:32:5
   |
LL |     foo(&mut "aaa".to_owned());
   |     ^^^^---------------------^
   |         |
   |         expected `String`, found `&mut String`
   |
note: function defined here
  --> $DIR/deref-suggestion.rs:5:4
   |
LL | fn foo(_: String) {}
   |    ^^^ ---------
help: provide an argument of the correct type
   |
LL |     foo({String});
   |     ^^^^^^^^^^^^^

error[E0308]: arguments to this function are incorrect
  --> $DIR/deref-suggestion.rs:34:5
   |
LL |     ($x:expr) => { &$x }
   |                    --- expected `u32`, found `&{integer}`
...
LL |     foo3(borrow!(0));
   |     ^^^^^^^^^^^^^^^^
   |
note: function defined here
  --> $DIR/deref-suggestion.rs:12:4
   |
LL | fn foo3(_: u32) {}
   |    ^^^^ ------
help: provide an argument of the correct type
   |
LL |     foo3({u32});
   |     ^^^^^^^^^^^

error[E0308]: mismatched types
  --> $DIR/deref-suggestion.rs:36:5
   |
LL |     assert_eq!(3i32, &3i32);
   |     ^^^^^^^^^^^^^^^^^^^^^^^^ expected `i32`, found `&i32`
   |
   = note: this error originates in a macro (in Nightly builds, run with -Z macro-backtrace for more info)

error[E0308]: mismatched types
  --> $DIR/deref-suggestion.rs:39:17
   |
LL |     let s = S { u };
   |                 ^
   |                 |
   |                 expected `&u32`, found integer
   |                 help: consider borrowing here: `u: &u`

error[E0308]: mismatched types
  --> $DIR/deref-suggestion.rs:41:20
   |
LL |     let s = S { u: u };
   |                    ^
   |                    |
   |                    expected `&u32`, found integer
   |                    help: consider borrowing here: `&u`

error[E0308]: mismatched types
  --> $DIR/deref-suggestion.rs:44:17
   |
LL |     let r = R { i };
   |                 ^
   |                 |
   |                 expected `u32`, found `&{integer}`
   |                 help: consider dereferencing the borrow: `i: *i`

error[E0308]: mismatched types
  --> $DIR/deref-suggestion.rs:46:20
   |
LL |     let r = R { i: i };
   |                    ^
   |                    |
   |                    expected `u32`, found `&{integer}`
   |                    help: consider dereferencing the borrow: `*i`

error: aborting due to 10 previous errors

For more information about this error, try `rustc --explain E0308`.
